### PR TITLE
VerifyOrigin post

### PR DESCRIPTION
**Pull request by danmcclain**

### CODE REVIEW

File: source/posts/2016-04-01-announcing-verify-origin.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+social: true
+summary: "Securing your API without a CSRF token"
+published: true
+tags: phoenix, security
+---
+
+In web applications, if you are using cookies for storing authentication
+state, you should be worrying about [Cross Site Request Forgery][csrf]
+(CSRF). In short, when using cookies, other sites can create an HTML
+form the submits to your backend. When your browser submits that form,
```
Comment (rsocci) [inline source/posts/2016-04-01-announcing-verify-origin.md]: `the` -> `that` submits to your backend

File: source/posts/2016-04-01-announcing-verify-origin.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+it will attach any cookies it has that are associated with the URL that
+the form is submitted to. Without any steps taken to mitigate this, a
+malicious third party could create a form on their site that performs an
+action on application, maybe as harmless as following them on a site, or
+as malicious as deleting your account, or transferring funds. [OWASP has
```
Comment (rsocci) [inline source/posts/2016-04-01-announcing-verify-origin.md]: no comma after `account`

File: source/posts/2016-04-01-announcing-verify-origin.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+malicious forms will be lacking this token and will be rejected. [More
+details on token based mitigation are covered in this
+OWASP page][owasp-csrf-mitigation].
+
+## What about Single Page applications? Is a JSON only API vulnerable?
+
+CSRF tokens work well for server side rendered applications (SSR), since
+the HTML generated on the server can securely include the token that is in the
+session cookie. But what about single page application like Ember? Can
+you craft an HTML from to submit a valid request to a JSON based API?
```
Comment (rsocci) [inline source/posts/2016-04-01-announcing-verify-origin.md]: HTML `form`

File: source/posts/2016-04-01-announcing-verify-origin.md (hunk 1 above)
Comment (danmcclain) [inline source/posts/2016-04-01-announcing-verify-origin.md]: I'm referencing the form in the previous sentence

File: source/posts/2016-04-01-announcing-verify-origin.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+## Places where CSRF prevention fails
+
+No matter how good your CSRF prevention is, there are other vectors that
+will open your application to malicious requests. If your session cookie
+is not `HTTPOnly` ([details][httponly]), and/or you accidentally allow malicious JavaScript to
+execute on your site, you have a couple of problems. When not using
+`HTTPOnly` cookies, your cookies can be retrieved via JavaScript, and if
+you have malicious JavaScript on your site, someone could siphon off all
+your session cookies. At that point, this third party could forge any
+header it needs to and include the session cookie to automate requests
```
Comment (rsocci) [inline source/posts/2016-04-01-announcing-verify-origin.md]: it needs ~~to~~ and

File: source/posts/2016-04-01-announcing-verify-origin.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+[VerifyOrigin][verify-origin] to
+[hex.pm](https://hex.pm/packages/verify_origin). VerifyOrigin
+provides a `Plug` you can add to your applications pipeline to only
+allow requests from `Origin`s you expect:
+
+```elixir
+plug VerifyOrigin, ["https://example.com"]
+```
+
+The plug expects list of valid `Origin` URLs. If the request does not
```
Comment (rsocci) [inline source/posts/2016-04-01-announcing-verify-origin.md]: expects **a** list of valid

File: source/posts/2016-04-01-announcing-verify-origin.md (hunk 1 above)
Comment (rsocci) [inline source/posts/2016-04-01-announcing-verify-origin.md]: It's "form that submits" vs "form the submits" @danmcclain

**Comment by rsocci** [comment]
nice plug @danmcclain! :tada:

File: source/posts/2016-04-01-announcing-verify-origin.md (hunk 1 above)
Comment (danmcclain) [inline source/posts/2016-04-01-announcing-verify-origin.md]: Ah, was looking at the end of the line not the beginning, sorry!

File: source/posts/2016-04-01-announcing-verify-origin.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+is not `HTTPOnly` ([details][httponly]), and/or you accidentally allow malicious JavaScript to
+execute on your site, you have a couple of problems. When not using
+`HTTPOnly` cookies, your cookies can be retrieved via JavaScript, and if
+you have malicious JavaScript on your site, someone could siphon off all
+your session cookies. At that point, this third party could forge any
+header it needs to, and then include the session cookie to automate requests
+against your server, so it could bypass the `Origin` check. They could
+also replace the cookie in their browser with another person's cookie
+that they stole, and use your site as that user, bypassing CSRF Tokens.
+If your cookie is marked as `HTTPOnly`, the can still use malicious
```
Comment (duggiefresh) [inline source/posts/2016-04-01-announcing-verify-origin.md]: `they can still use malicious...`

**Comment by chrismccord** [comment]
:shipit:

**Pull request merged**